bee-storage/bee-storage-rocksdb/src/access/fetch.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    column_families::*,
+    storage::{Storage, StorageBackend},
+    system::System,
+};
+
+use bee_common::packable::Packable;
+use bee_ledger::{
+    snapshot::info::SnapshotInfo,
+    types::{Balance, ConsumedOutput, CreatedOutput, LedgerIndex, OutputDiff, Receipt, TreasuryOutput},
+};
+use bee_message::{
+    address::{Address, Ed25519Address, ED25519_ADDRESS_LENGTH},
+    milestone::{Milestone, MilestoneIndex},
+    output::{OutputId, OUTPUT_ID_LENGTH},
+    payload::indexation::{PaddedIndex, INDEXATION_PADDED_INDEX_LENGTH},
+    Message, MessageId, MESSAGE_ID_LENGTH,
+};
+use bee_storage::access::Fetch;
+use bee_tangle::{
+    metadata::MessageMetadata, solid_entry_point::SolidEntryPoint, unconfirmed_message::UnconfirmedMessage,
+};
+
+use std::convert::{TryFrom, TryInto};
+
+#[async_trait::async_trait]
+impl Fetch<u8, System> for Storage {
+    async fn fetch(&self, key: &u8) -> Result<Option<System>, <Self as StorageBackend>::Error> {
+        Ok(self
+            .inner
+            .get_cf(self.cf_handle(CF_SYSTEM)?, [*key])?
+            // Unpacking from storage is fine.
+            .map(|v| System::unpack_unchecked(&mut v.as_slice()).unwrap()))
+    }
+}
+
+#[async_trait::async_trait]
+impl Fetch<MessageId, Message> for Storage {
+    async fn fetch(&self, message_id: &MessageId) -> Result<Option<Message>, <Self as StorageBackend>::Error> {
+        Ok(self
+            .inner
+            .get_cf(self.cf_handle(CF_MESSAGE_ID_TO_MESSAGE)?, message_id)?
+            // Unpacking from storage is fine.
+            .map(|v| Message::unpack_unchecked(&mut v.as_slice()).unwrap()))
+    }
+}
+
+#[async_trait::async_trait]
+impl Fetch<MessageId, MessageMetadata> for Storage {
+    async fn fetch(&self, message_id: &MessageId) -> Result<Option<MessageMetadata>, <Self as StorageBackend>::Error> {
+        Ok(self
+            .inner
+            .get_cf(self.cf_handle(CF_MESSAGE_ID_TO_METADATA)?, message_id)?
+            // Unpacking from storage is fine.
+            .map(|v| MessageMetadata::unpack_unchecked(&mut v.as_slice()).unwrap()))
+    }
+}
+
+#[async_trait::async_trait]
+impl Fetch<MessageId, Vec<MessageId>> for Storage {
+    async fn fetch(&self, parent: &MessageId) -> Result<Option<Vec<MessageId>>, <Self as StorageBackend>::Error> {
+        Ok(Some(
+            self.inner
+                .prefix_iterator_cf(self.cf_handle(CF_MESSAGE_ID_TO_MESSAGE_ID)?, parent)
+                .map(|(key, _)| {
+                    let (_, child) = key.split_at(MESSAGE_ID_LENGTH);
```

Comment:
is there no way to do `key[..MESSAGE_ID_LENGTH]`?

Reply:
That's not really equivalent. We'd need something like `key[key.len() - MESSAGE_ID_LENGTH..]`.